### PR TITLE
state-utils: Extract extendAction

### PR DESCRIPTION
**Pull request by sarayourfriend**
#### Changes proposed in this Pull Request

* Extract `extendAction` into `state-utils`.

#### Testing instructions

This will test the `withoutNotice` function which uses `extendAction`.
* Open Pages in Calypso.
* Publish a new page, update and save an existing page, trash a page and restore a page from a previous version, and verify you don't see Calypso notices when you do those.

I couldn't find an example to test where we use `extendAction` over a `thunk` though 🤔 I think it should be fine as I didn't change any of the logic of the function, just the types.

### CODE REVIEW

File: client/state/README.md [hunk 1]
```diff
@@ -49,28 +49,6 @@ All the application information and data in Calypso should go through this data
 
 `state/utils.js` contains a number of helper utilities you may find useful in implementing your state subtree:
 
-### extendAction( action, data )
```
Comment (tyxla) [inline client/state/README.md:52]: Now that we're moving these to packages, should we update the documentation to say that some of the state utilities are in the `state-utils` package, and we're in the process of moving the rest there?

**Review by tyxla — COMMENTED** [review]
Sorry to set this back a bit, but could we split the extraction and conversion to TS to separate PRs? It seems those should happen separately.

**Review by tyxla — APPROVED** [review]
Things are looking great here!

I pushed a critical commit to fix the `extendAction` export, and several more nitty ones, but other than that this looks and works ace! 

Thanks, @sarayourfriend!

**Comment by jsnajdr** [comment]
Some observations: `extendAction` is a special-purpose utility that's used only in the data layer Redux code. So it might be a private utility of that library rather than a general-purpose helper.

Other usages can be removed quite easily, with reduction in complexity. See #50940.

It's almost certain that the usage in data layer doesn't require supporting thunks. `extendAction` could get way simpler if we removed that support and if it was a simple "merge a `meta` property into an action object" function.

**Comment by tyxla** [comment]
> Some observations: `extendAction` is a special-purpose utility that's used only in the data layer Redux code. So it might be a private utility of that library rather than a general-purpose helper.
> 
> Other usages can be removed quite easily, with reduction in complexity. See #50940.
> 
> It's almost certain that the usage in data layer doesn't require supporting thunks. `extendAction` could get way simpler if we removed that support and if it was a simple "merge a `meta` property into an action object" function.

Good call, @jsnajdr! While not blocking this PR at all, those are some good ideas for follow-up PRs.

**Pull request merged**